Implement Vision Subsystem

Geotricity opened the pull request:
- Create Vision class that gets frames from the camera
- Create GripPipeline class with methods to process images from the camera and isolate the target
- Add VisionAim command that aims the robot at the target
- Link VisionAim to button 9 on the controller
- Add Vision subsystem to Robot
- Add initCamera() to Robot.initialize()

Geotricity committed:
Implement Vision Subsystem

- Create Vision class that gets frames from the camera
- Create GripPipeline class with method to process images from the camera and isolate the target
- Add VisionAim command that aims the robot at the target
- Link VisionAim to button 9 on the controller
- Add Vision subsystem to Robot
- Add initCamera() to Robot.initialize()

bricktrooper commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -24,26 +24,29 @@
 public class Robot
 {
   // SUBSYSTEMS //
+  public Driver driver;
```
Maybe put `driver` under a separate comment called `// CONTROLS //` or something like that.

bricktrooper replied:
Also we should instantiate the operator here as well.  Can be a separate PR if you want.

bricktrooper commented on src/main/java/frc/robot/subsystems/Vision.java:
```diff
@@ -0,0 +1,88 @@
+  This method makes the robot turn in place 
+  to face the target
+  ===================================== */
+  public void aimAtTarget() {
+    double xCentre;
+    synchronized (imgLock) {
+        xCentre = this.xCentre;
+    }
+    double turn = xCentre - (WIDTH / 2); // find difference between the target and where the robot is pointed
+    robot.drivetrain.moveMotors(0, turn * TURN_SPEED);
+  }
```
Generally you shouldn't pass `robot` to subsystems because they **are** the robot.  In this case, `Vision` is essentially the code that controls the camera and does any related processing.  You don't want subsystems to hijack other subsystems (in this case, `drivetrain`) because that will confuse the motors if you ever call `vision.aimAtTarget()` and a `drivetrain.moveMotors()`  at the same time.  The way to properly take control of another subsystem is via a command.  In this case you would turn the `drivetrain` in `VisionAim` and that way you can enforce that the scheduler will grant access to the drivetrain using `addRequirements()`.  The `Vision` subsystem should only control the camera and do camera-related stuff but nothing else.  So instead of directly controlling of the drivetrain here, you could just have the vision subsystem return the turn value that is needed for alignment and then pass that to the drivetrain in your command.

Geotricity replied:
That makes sense, I'll make sure to do that in the future.

danielzyy replied:
Yes, call `drivetrain.moveMotors()` inside the `VisionAim` command instead of in subsystem.

bricktrooper commented:
I don't have much experience in the vision processing itself, but this makes sense to me.  I'll let @danielzyy and @droid-101 look at this since they have done this before.

danielzyy commented on src/main/java/frc/robot/subsystems/Vision.java:
```diff
@@ -0,0 +1,88 @@
+    visionThread = new VisionThread(cam, new GripPipeline(), pipeline -> {
+      
+      if (!pipeline.filterContoursOutput().isEmpty()) {
+        Rect r = Imgproc.boundingRect(pipeline.filterContoursOutput().get(0)); //Gets bounding rectangle around the target
```
This is assuming you only have one bounding rectangle, what happens if you see multiple? Is that possible in the competition (eg. multiple pieces of tape in camera frame). If so, consider if you need to sort the bounding rectangles by area to get the closest target rectangle to track, so that you are not accidentally aligning on a further piece of tape. I remember I created a comparator function to just sort the rectangles by area, there may be a better way to sort them though: https://github.com/team4015/FRC4015-2020/blob/29296e80394bf4ff7bc53c57a3207a89d7921e2d/src/main/java/frc/robot/commands/auto/AutoAlign.java#L85

danielzyy replied:
It was also nice to visualize the bounding box on the camera frame so that you can visually verify that it is working properly. Just draw a rectangle on the input frame to represent the bounding box and then output it to a camera stream so that you can see it on the dashboard.

Geotricity replied:
I've added a method that finds the biggest contour after looking at yours, and I've outputted that to the dashboard, but I haven't tested it yet. Thanks for pointing this out, I didn't think about this at all.

Geotricity committed:
Fix Suggestions

- Instantiates Operator in Robot
- Makes the VisonAim command move the drivetrain instead of the Vision class itself
- (Potential bug fix) Scaled xCentre to match the resolution of the camera

danielzyy commented on src/main/java/frc/robot/subsystems/Vision.java:
```diff
@@ -0,0 +1,88 @@
+        xCentre = this.xCentre;
+    }
+    double turn = xCentre - (WIDTH / 2); // find difference between the target and where the robot is pointed
+    robot.drivetrain.moveMotors(0, turn * TURN_SPEED);
```
When testing with low speeds, this could work to stop the robot from turning when perfectly aligned, but at higher turn speeds, due to the camera input delay, there is a good chance you will overshoot the target and end up swerving back the other way to correct it, leading to repeated oscillations(you will never get exactly on-target). A way to solve this is to add a threshold in the `turn` value for the alignment to be "good enough" and stop the robot from turning earlier instead of stopping when it is exactly on-target(to compensate for the camera lag). Another thing you can add (when you have time) is to make the turning speed proportional (Think of a PID control loop but just the proportional part), so that when the robot is more misaligned, it will turn faster, and when it gets closer to the target, it will turn slower. This is optional however since it would work well with just a fixed speed and threshold too.

Geotricity replied:
I added a threshold to the aiming. The speed is already proportional to the gap between the centre and the target. Do you think a full PID would be useful here, or is this good enough?

danielzyy replied:
Just using proportional control with a threshold is enough for the first competition. If you have time afterwards and you see that it can be improved, then you can maybe try implementing PI loop, probably don't need a full PID. Tuning the PID values will also be a pain so only work on it if it is critical and there is nothing else needed to be done.

Geotricity replied:
Okay, thanks.

danielzyy commented on src/main/java/frc/robot/commands/vision/VisionAim.java:
```diff
@@ -0,0 +1,65 @@
+  @Override
+  public void execute()
+  {
+    robot.vision.aimAtTarget();
```
Drive the drivetrain motors here, not in the subsystem, you can get the turn values as a return from the subsystem like Kyle mentioned.

danielzyy reviewed (commented):
I know this might be the basic vision implementation so far, but added some additional considerations just to make sure you are aware of them.

Geotricity committed:
Merge branch 'master' into vision

Geotricity committed:
Merge branch 'master' into vision

Geotricity committed:
Fix Suggestions

- Finds the biggest contour instead of the first
- Outputs camera to drive station
- Added threshold to auto aim

Geotricity committed:
Initalize xCentre

- Set xCentre to WIDTH/2.0 initially to avoid errors

Geotricity committed:
Merge branch 'master' into vision

Geotricity committed:
Merge branch 'master' into vision

Geotricity committed:
Merge branch 'vision' of https://github.com/team4015/FRC-Robot-2022-Rapid-React into vision

Geotricity committed:
Instantiate controls after subsystems

-  instantiating the controls before the subsystems causes an error. This is because Operator calls commands, which call subsystem methods. But the subsystems haven't been instantiated yet, and we get a null error

Geotricity commented:
It would be really useful if this PR could get approved or suggestions could be given soon so that I can work on the camera in a different branch after school today.

danielzyy reviewed (approved):
Looks good, `TURN_SPEED` will need to be tuned when testing auto-alignment.
